Implement the coap-mesage traits on request

chrysn opened the pull request:
This implements the generic CoAP manipulation traits of the [coap-message](https://crates.io/crates/coap-message) crate for coap-lite.

This allows writing CoAP applications in a more portable way, as a handler can parse a request and render a response independently of the underlying CoAP implementation used, be it coap-lite, coap-rs, RIOT-OS's [gcoap through the Rust wrappers](https://crates.io/crates/riot-wrappers) or even [CoAP-over-WebSockets (and BLE) in the browser](https://crates.io/crates/verdigris), or a WIP [CoAP server based on embedded-nal](https://gitlab.com/chrysn/embedded-nal-minimal-coapserver).

Implementation is done here rather than in the traits crate as a) I prefer trait-only crates to not ship their own implementations (with the exception of general implementations based on primitives), and b) the interior modification of CoAP options that's part of the MutableWritableMessage trait is more efficiently implemented by accessing the internal options data (which is thus promoted to `pub(crate)`).

During the implementation I noted that option numbers are internally stored as usize, which opens up panic-paths as CoAP options are by design 16-bit numbers. I didn't try to change anything because they're part of the public interface (and because I prefer to keep PRs to the point), but if that ever changes, the `impl OptionNumber for CoapOption` auxiliary impls can go away and make the whole thing slimmer.

I'll follow up with an example to be added to the PR, but writing that showed that the [MutableWritableMessage](https://docs.rs/coap-message/0.1.0/coap_message/trait.MutableWritableMessage.html) interface behaves a bit odd for messages that do not have preallocated buffers. This comes as no surprise (given that, unlike the other message interfaces, it is documented to be a bit of a mess still), and is a good datum for enhancing if not the interface right now at least the implementations that are to be used for the example.

martindisch commented:
That's very nice to see! Especially since I noticed your efforts for `coap-message` and thought it was kind of neat, way back when I was actively working on this.
I'll give this the attention it deserves when I have time tomorrow or on the weekend at the latest.

chrysn committed:
coap-message: Implement marker traits already satisfied by construction

chrysn committed:
coap-message: Bump to 0.1.1 to make payload_mut more usable

Using coap-message with coap-lite has shown its MutableWritableMessage's
limitations when used with dynamically allocated payloads; the new
added method fixes this.

martindisch commented:
Looks good to me! For the usize problem, are you referring to cases where usize is less than u16, so 8-bit MCUs? Or are there other potential issues?

I'm happy to wait for the example, but it's not a dealbreaker for me if there isn't one. If you're short on time and would like this merged without it, just let me know.

martindisch commented:
There is one strange issue, probably caused by the auxiliary impls you mentioned. A test doesn't compile when your new feature is active, can be fixed with
```diff
diff --git a/src/packet.rs b/src/packet.rs
index eb75bd0..82e8929 100644
--- a/src/packet.rs
+++ b/src/packet.rs
@@ -675,7 +675,7 @@ mod test {
 
     #[test]
     fn option() {
-        for i in 0..512 {
+        for i in 0_usize..512 {
             match CoapOption::try_from(i) {
                 Ok(o) => assert_eq!(i, o.into()),
                 _ => (),
```
I would've pushed that to the PR right away, but it looks like you didn't activate the option giving me access to your branch.

Regarding the u16 problem, I'm totally fine with breaking the public interface if it serves a purpose, that's what we have SemVer for after all. I gave it a shot at https://github.com/martindisch/coap-lite/tree/16-bit-option. If that makes sense to you, I can merge it to master and let you rebase on top of it, so we can throw out the additional cruft right in this PR.

chrysn commented:
On the usize vs u16: No (I don't think an MCU with usize = u8 really makes sense), it's just about possibly illegal values cropping up in the API that'd need fallible conversions, but the conversion is infallible because *all* options have u16 numbers, so it has a potential to panic which the compiler can't optimize away b/c it can't be sure that nobody ever puts a large number into the usize. It's not a big issue, just a wasted potential for optimization (2 bytes on every option, and the exception messages / code paths that are in the flash forever dead).

However, if now is a good time to make that semver bump, then yes, go for it -- and I'll rebase and adjust the branch. (Saves ten lines or so too.)

Yes, the test breaks because there impls add try-from and into u16 for the options -- could be done with newtypes too but I think it makes sense the way it is. Pushed your fix (along with two further additions that turned out to make sense while working on the examples).

martindisch commented:
Thanks for the explanation. Regarding additional code paths and potential for panics, there are probably some other things that could be optimized a bit. I didn't change much of the original code when I forked it to make it run on no_std.

Done. Feel free to mention anything else if something jumps out at you.

chrysn committed:
Add example using coap-message and -handler

chrysn committed:
Implement the coap-mesage traits on request

chrysn commented:
There's now examples in, and all rebased to latest master with 16bit option numbers (reducing the code a bit because the from/to 16bit is now already implemented).

Usage example attached by attachment because for some reasons github comments, even though being HTML based, don't allow my text colors:
![coapclient](https://user-images.githubusercontent.com/442326/107145125-e7569780-693f-11eb-8a95-9643cc165e74.png)

(The [::ffff:127.0.0.1] addresses are due to a bug in aiocoap that surfaces with IPv4 and sends the IP address in a Uri-Host header that is not recognized by the server right now.)

martindisch commented:
Thank you, I'll merge and release the new version soon.
